crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//!     circuit
+//!         .add_source(CsvSource::<_, Record, isize, OrdZSet<_, _>>::from_csv_reader(reader()?))
+//!         .inspect(|data| println!("{}", data.weighted_count()));
+//!     Ok(())
+//! }
+//! ```
+//!
+//! There's still nothing that happens if you run it now, because we're
+//! still not running the circuit.  To run it a single step, we insert a
+//! call to [`CircuitHandle::step`] near the end of `main().`
+//!
+//! Now, if we run it, it prints `3961`, the number of records in
+//! `vaccinations.csv`.  That's because our program reads an entire CSV file and
+//! feeds it as input in a single step.  That means that running for more steps
+//! wouldn't make a difference.  That's not a normal use case for DBSP but,
+//! arguably, it's a reasonable setup for a tutorial.
+//!
+//! The full program is in `tutorial3.rs`.
+//!
+//! ## Processing
+//!
+//! We've got the data now.  Let's write a few queries on it to see how
+//! that would work.
+//!
+//! ### Selecting a subset
+//!
+//! This is a lot of data.  Suppose we want to pick out a subset of the
+//! records.  We can use the [`FilterMap`] trait implemented for [`Stream`] to
+//! do that.  For example, we can take just the records for locations in
+//! the United Kingdom:
+//!
+//! ```rust
+//! # use anyhow::Result;
+//! # use csv;
+//! use dbsp::{
+//!     operator::{
+//!         time_series::{RelOffset, RelRange},
+//!         FilterMap,
+//!     },
+//!     CollectionHandle, IndexedZSet, OrdZSet, RootCircuit, ZSet,
+//! };
+//! # use serde::Deserialize;
+//! # use size_of::SizeOf;
+//! # use time::Date;
+//!
+//! # #[derive(Clone, Debug, Deserialize, Eq, PartialEq, Ord, PartialOrd, Hash, SizeOf)]
+//! # struct Record {
+//! #     location: String,
+//! #     date: Date,
+//! #     daily_vaccinations: Option<u64>,
+//! # }
+//!
+//! fn build_circuit(circuit: &mut RootCircuit) -> Result<CollectionHandle<Record, isize>> {
+//!     let (input_stream, input_handle) = circuit.add_input_zset::<Record, isize>();
+//!     input_stream
+//!         .filter(|r| {
+//!             r.location == "England"
+//!                 || r.location == "Northern Ireland"
+//!                 || r.location == "Scotland"
+//!                 || r.location == "Wales"
+//!         })
+//!         .inspect(move |record: &OrdZSet<Record, isize>| {
+//!             println!("{}", record.weighted_count());
+//!         });
+//!     Ok(input_handle)
+//! }
+//! ```
+//!
+//! This prints `3083`, showing that we did select a subset of the 3,961
+//! total records.
+//!
+//! The full program is in `tutorial4.rs`.
+//!
+//! ### Aggregating by month
+//!
+//! 3,083 records is still a lot to look at.  There's so much because we've got
+//! years of daily data.  Let's aggregate daily vaccinations into months, to get
+//! monthly vaccinations.  DBSP has several forms of aggregation.  All of them
+//! work with "indexed Z-sets" ([`IndexedZSet`]), which are
```

Comment:
I think that sql group-by is a better analogy. they are zsets grouped by a key.

Reply:
in general I think showing the equivalent SQL is very useful for people who know SQL.